SafeExamBrowser.SystemComponents/VirtualMachineDetector.cs

```diff
@@ -37,27 +39,136 @@ public VirtualMachineDetector(ILogger logger, ISystemInfo systemInfo)
 			this.systemInfo = systemInfo;
 		}
 
-		public bool IsVirtualMachine()
+		private bool IsVirtualSystemInfo(string biosInfo, string manufacturer, string model)
 		{
-			var biosInfo = systemInfo.BiosInfo.ToLower();
-			var isVirtualMachine = false;
-			var macAddress = systemInfo.MacAddress;
-			var manufacturer = systemInfo.Manufacturer.ToLower();
-			var model = systemInfo.Model.ToLower();
-			var devices = systemInfo.PlugAndPlayDeviceIds;
+			bool isVirtualMachine = false;
+
+			biosInfo = biosInfo.ToLower();
+			manufacturer = manufacturer.ToLower();
+			model = model.ToLower();
 
 			isVirtualMachine |= biosInfo.Contains("hyper-v");
 			isVirtualMachine |= biosInfo.Contains("virtualbox");
 			isVirtualMachine |= biosInfo.Contains("vmware");
+			isVirtualMachine |= biosInfo.Contains("ovmf");
+			isVirtualMachine |= biosInfo.Contains("edk ii unknown");  // qemu
 			isVirtualMachine |= manufacturer.Contains("microsoft corporation") && !model.Contains("surface");
 			isVirtualMachine |= manufacturer.Contains("parallels software");
 			isVirtualMachine |= manufacturer.Contains("qemu");
 			isVirtualMachine |= manufacturer.Contains("vmware");
 			isVirtualMachine |= model.Contains("virtualbox");
+			isVirtualMachine |= model.Contains("Q35 +");  // qemu
+
+			return isVirtualMachine;
+		}
+
+		private bool IsVirtualRegistry()
+		{
+			bool isVirtualMachine = false;
+
+			// check historic hardware profiles
+			RegistryKey hardwareConfig = Microsoft.Win32.Registry.LocalMachine.OpenSubKey("SYSTEM\\HardwareConfig");
```

Comment:
Please access the registry via the already existing `IRegistry` implementation: https://github.com/SafeExamBrowser/seb-win-refactoring/blob/master/SafeExamBrowser.SystemComponents/Registry/Registry.cs. You may need to extend the `IRegistry` interface, if you'd need help or assistance on why and how to do that, I'd be happy to help.

Reply:
I indeed noticed the IRegistry class - however, I noted that in order to get the subkeys of a key (`.GetSubKeyNames()`), I needed access to the `RegistryKey` objects, which I could only retrieve using `.OpenSubKey()` on registry hive objects like `Registry.LocalMachine`, `Registry.CurrentUser`, etc. objects. The issue lies in that I can't figure out how to dynamically access hive objects like LocalMachine without making an ugly switch statement and hence making a breaking change to the interface.

Should I do the latter or do you know any tricks to circumvent this?

Reply:
I think you'll need to extend the interface in order to retrieve e.g. the subkeys of a key, so e.g. add a method `bool TryGetSubKeys(string key, out IEnumerable<string> subKeys)` to the IRegistry interace.

If you'd like, we can have a look at this together or you can keep that point open and I'd implement it as an example which you then could reference in the future?

Reply:
I have extended the interface as desired, and added the following IRegistry functions: `bool TryGetNames(string key, out IEnumerable<string> names)` and `public bool TryGetSubKeys(string key, out IEnumerable<string> subKeys)`, both of which are wrappers.

By the way, is it a SEB convention to use negative return values instead of exceptions for error handling?

Reply:
>By the way, is it a SEB convention to use negative return values instead of exceptions for error handling?

I guess that's an issue open for debate, but I've decided to apply the try-pattern where meaningful instead of using exceptions to handle control flow. I vaguely remember some code wizard proclaiming that exceptions should never be used for control flow, but I guess in the end that's a matter of personal preference and always depending on the actual context of a piece of code etc.